# Date Night Movie

In this assignment we are going to use pandas to figure out - What's the best **date-night movie**?

This assignment is going to use
- Joining
- Groupby
- Sorting



In [2]:
import os
import pandas as pd

# Read in the movie data: `pd.read_table`

In [89]:
def get_movie_data():
    
    unames = ['user_id','gender','age','occupation','zip']
    users = pd.read_table(os.path.join('../data','users.dat',), 
                          sep='::', header=None, names=unames, encoding='latin1')
    
    rnames = ['user_id', 'movie_id', 'rating', 'timestamp']
    ratings = pd.read_table(os.path.join('../data', 'ratings.dat'), 
                            sep='::', header=None, names=rnames, encoding='latin1').sort_values("timestamp")
    ratings["timestamp"] = pd.to_datetime(ratings["timestamp"], unit='s')

    
    mnames = ['movie_id', 'title','genres']
    movies = pd.read_table(os.path.join('../data', 'movies.dat'), 
                           sep='::', header=None, names=mnames, encoding='latin1')

    return users, ratings, movies

In [90]:
users, ratings, movies = get_movie_data()

C:\Users\varsh\AppData\Local\Temp\ipykernel_42984\2990020759.py:4: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  users = pd.read_table(os.path.join('../data','users.dat',),
C:\Users\varsh\AppData\Local\Temp\ipykernel_42984\2990020759.py:8: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  ratings = pd.read_table(os.path.join('../data', 'ratings.dat'),
C:\Users\varsh\AppData\Local\Temp\ipykernel_42984\2990020759.py:14: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avo

In [91]:
users.head()

user_id gender  age  occupation    zip
0        1      F    1          10  48067
1        2      M   56          16  70072
2        3      M   25          15  55117
3        4      M   45           7  02460
4        5      M   25          20  55455

In [92]:
users.shape

(6040, 5)

In [93]:
ratings.head()

user_id  movie_id  rating           timestamp
1000138     6040       858       4 2000-04-25 23:05:32
1000153     6040      2384       4 2000-04-25 23:05:54
999873      6040       593       5 2000-04-25 23:05:54
1000007     6040      1961       4 2000-04-25 23:06:17
1000192     6040      2019       5 2000-04-25 23:06:17

In [94]:
ratings.shape

(1000209, 4)

In [95]:
movies.head()

movie_id                               title                        genres
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

In [96]:
movies.shape

(3883, 3)

##### Clean up the `movies`

- Get the `year`
- Shorten the `title`

In [97]:
tmp = movies.title.str.extract('(.*) \(([0-9]+)\)')
tmp.apply(lambda x:x[0] if len(x) > 0 else None)
tmp.apply(lambda x: x[0][:40] if len(x) > 0 else None)

0    Toy Story
1         1995
dtype: object

In [98]:
movies['year'] = tmp[1]
movies['short_title'] = tmp[0]

In [99]:
movies.head()

movie_id                               title                        genres  \
0         1                    Toy Story (1995)   Animation|Children's|Comedy   
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy   
2         3             Grumpier Old Men (1995)                Comedy|Romance   
3         4            Waiting to Exhale (1995)                  Comedy|Drama   
4         5  Father of the Bride Part II (1995)                        Comedy   

   year                  short_title  
0  1995                    Toy Story  
1  1995                      Jumanji  
2  1995             Grumpier Old Men  
3  1995            Waiting to Exhale  
4  1995  Father of the Bride Part II

# Join the tables with `pd.merge` (20 pts)

merge users and ratings with user_id as common colum

In [100]:
users_ratings = pd.merge(users, ratings, on=['user_id'], how='inner')
users_ratings

user_id gender  age  occupation    zip  movie_id  rating  \
0              1      F    1          10  48067      3186       4   
1              1      F    1          10  48067      1721       4   
2              1      F    1          10  48067      1270       5   
3              1      F    1          10  48067      1022       5   
4              1      F    1          10  48067      2340       3   
...          ...    ...  ...         ...    ...       ...     ...   
1000204     6040      M   25           6  11106      2917       4   
1000205     6040      M   25           6  11106      1921       4   
1000206     6040      M   25           6  11106      1784       3   
1000207     6040      M   25           6  11106       161       3   
1000208     6040      M   25           6  11106      1221       4   

                  timestamp  
0       2000-12-31 22:00:19  
1       2000-12-31 22:00:55  
2       2000-12-31 22:00:55  
3       2000-12-31 22:00:55  
4       2000-12-31 22:01:43  
...                     ...  
1000204 2001-08-10 14:40:29  
1000205 2001-08-10 14:41:04  
1000206 2001-08-10 14:41:04  
1000207 2001-08-10 14:41:26  
1000208 2001-08-20 13:44:15  

[1000209 rows x 8 columns]

merge users_ratings and movies with movie_id as common colum

In [101]:
users_ratings_movies = pd.merge(users_ratings, movies, on=['movie_id'], how='inner')
users_ratings_movies

user_id gender  age  occupation    zip  movie_id  rating  \
0              1      F    1          10  48067      3186       4   
1              8      M   25          12  11413      3186       4   
2             33      M   45           3  55421      3186       3   
3             90      M   56          13  85749      3186       3   
4             99      F    1          10  19390      3186       3   
...          ...    ...  ...         ...    ...       ...     ...   
1000204     5949      M   18          17  47901      2198       5   
1000205     5675      M   35          14  30030      2703       3   
1000206     5780      M   18          17  92886      2845       1   
1000207     5851      F   18          20  55410      3607       5   
1000208     5938      M   25           1  35401      2909       4   

                  timestamp                                        title  \
0       2000-12-31 22:00:19                     Girl, Interrupted (1999)   
1       2000-12-31 02:47:32                     Girl, Interrupted (1999)   
2       2000-12-29 17:11:29                     Girl, Interrupted (1999)   
3       2001-06-30 04:30:17                     Girl, Interrupted (1999)   
4       2000-12-23 21:53:02                     Girl, Interrupted (1999)   
...                     ...                                          ...   
1000204 2000-05-20 18:13:21                           Modulations (1998)   
1000205 2000-12-05 15:11:56                        Broken Vessels (1998)   
1000206 2000-05-12 17:37:48                            White Boys (1999)   
1000207 2000-05-08 03:30:08                     One Little Indian (1973)   
1000208 2000-05-02 13:15:53  Five Wives, Three Secretaries and Me (1998)   

                       genres  year                           short_title  
0                       Drama  1999                     Girl, Interrupted  
1                       Drama  1999                     Girl, Interrupted  
2                       Drama  1999                     Girl, Interrupted  
3                       Drama  1999                     Girl, Interrupted  
4                       Drama  1999                     Girl, Interrupted  
...                       ...   ...                                   ...  
1000204           Documentary  1998                           Modulations  
1000205                 Drama  1998                        Broken Vessels  
1000206                 Drama  1999                            White Boys  
1000207  Comedy|Drama|Western  1973                     One Little Indian  
1000208           Documentary  1998  Five Wives, Three Secretaries and Me  

[1000209 rows x 12 columns]

### joining movies and ratings on movie_id and also setting index

In [113]:
movies_ratings = (ratings
                  .set_index("movie_id")
                  .join(movies.set_index("movie_id"),
                        how="left")
                 )
movies_ratings.head()

user_id  rating           timestamp             title  \
movie_id                                                          
1            6035       4 2000-04-26 01:34:09  Toy Story (1995)   
1            6032       4 2000-04-26 03:02:07  Toy Story (1995)   
1            6022       5 2000-04-26 13:29:23  Toy Story (1995)   
1            6021       3 2000-04-26 13:52:27  Toy Story (1995)   
1            6016       4 2000-04-26 19:52:30  Toy Story (1995)   

                               genres  year short_title  
movie_id                                                 
1         Animation|Children's|Comedy  1995   Toy Story  
1         Animation|Children's|Comedy  1995   Toy Story  
1         Animation|Children's|Comedy  1995   Toy Story  
1         Animation|Children's|Comedy  1995   Toy Story  
1         Animation|Children's|Comedy  1995   Toy Story

#### Creating seperate columns for genres by using pandas.get_dummies as they are all in one column seperated by pipe delimiter 
- and giving them boolean values 

In [118]:
genres_dummies = movies_ratings['genres'].str.get_dummies()
genres_dummies.head()

Action  Adventure  Animation  Children's  Comedy  Crime  \
movie_id                                                            
1              0          0          1           1       1      0   
1              0          0          1           1       1      0   
1              0          0          1           1       1      0   
1              0          0          1           1       1      0   
1              0          0          1           1       1      0   

          Documentary  Drama  Fantasy  Film-Noir  Horror  Musical  Mystery  \
movie_id                                                                     
1                   0      0        0          0       0        0        0   
1                   0      0        0          0       0        0        0   
1                   0      0        0          0       0        0        0   
1                   0      0        0          0       0        0        0   
1                   0      0        0          0       0        0        0   

          Romance  Sci-Fi  Thriller  War  Western  
movie_id                                           
1               0       0         0    0        0  
1               0       0         0    0        0  
1               0       0         0    0        0  
1               0       0         0    0        0  
1               0       0         0    0        0

#### Concatenate these dummies to the original movies_ratings data frame


In [120]:
full_movies_ratings = (pd.concat([movies_ratings, genres_dummies], axis=1)
                       .drop(["timestamp", "genres"], axis=1)
                )

full_movies_ratings.head()

user_id  rating             title  year short_title  Action  \
movie_id                                                                
1            6035       4  Toy Story (1995)  1995   Toy Story       0   
1            6032       4  Toy Story (1995)  1995   Toy Story       0   
1            6022       5  Toy Story (1995)  1995   Toy Story       0   
1            6021       3  Toy Story (1995)  1995   Toy Story       0   
1            6016       4  Toy Story (1995)  1995   Toy Story       0   

          Adventure  Animation  Children's  Comedy  ...  Fantasy  Film-Noir  \
movie_id                                            ...                       
1                 0          1           1       1  ...        0          0   
1                 0          1           1       1  ...        0          0   
1                 0          1           1       1  ...        0          0   
1                 0          1           1       1  ...        0          0   
1                 0          1           1       1  ...        0          0   

          Horror  Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
movie_id                                                                     
1              0        0        0        0       0         0    0        0  
1              0        0        0        0       0         0    0        0  
1              0        0        0        0       0         0    0        0  
1              0        0        0        0       0         0    0        0  
1              0        0        0        0       0         0    0        0  

[5 rows x 23 columns]

#### Resetting the index

In [121]:
full_movies_ratings.reset_index(inplace=True)

full_movies_ratings.head()

movie_id  user_id  rating             title  year short_title  Action  \
0         1     6035       4  Toy Story (1995)  1995   Toy Story       0   
1         1     6032       4  Toy Story (1995)  1995   Toy Story       0   
2         1     6022       5  Toy Story (1995)  1995   Toy Story       0   
3         1     6021       3  Toy Story (1995)  1995   Toy Story       0   
4         1     6016       4  Toy Story (1995)  1995   Toy Story       0   

   Adventure  Animation  Children's  ...  Fantasy  Film-Noir  Horror  Musical  \
0          0          1           1  ...        0          0       0        0   
1          0          1           1  ...        0          0       0        0   
2          0          1           1  ...        0          0       0        0   
3          0          1           1  ...        0          0       0        0   
4          0          1           1  ...        0          0       0        0   

   Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0       0         0    0        0  
1        0        0       0         0    0        0  
2        0        0       0         0    0        0  
3        0        0       0         0    0        0  
4        0        0       0         0    0        0  

[5 rows x 24 columns]

# What's the highest rated movie? (20 pts))

### Getting number of users and movies from the dataset.

In [103]:
user_ids = ratings.user_id.unique().tolist()
movie_ids = ratings.movie_id.unique().tolist()
print('Number of Users: {}'.format(len(user_ids)))
print('Number of Movies: {}'.format(len(movie_ids)))

Number of Users: 6040
Number of Movies: 3706


### Groupby movie_id

In [104]:
movie_id_group = users_ratings_movies.groupby(['movie_id','short_title'])['rating'].agg(['count','mean'])
movie_id_group 

count      mean
movie_id short_title                                 
1        Toy Story                     2077  4.146846
2        Jumanji                        701  3.201141
3        Grumpier Old Men               478  3.016736
4        Waiting to Exhale              170  2.729412
5        Father of the Bride Part II    296  3.006757
...                                     ...       ...
3948     Meet the Parents               862  3.635731
3949     Requiem for a Dream            304  4.115132
3950     Tigerland                       54  3.666667
3951     Two Family House                40  3.900000
3952     Contender, The                 388  3.780928

[3706 rows x 2 columns]

## Sorting 

#### Sorting by mean first and count for the movie ratings 

- sorting it to the descending order 
- Which gives an output of the highest mean of the movie ratings but rated by only three users which is **not statistically significant.**

In [105]:
movie_id_group.sort_values(by=['mean','count'], ascending=False)

count  mean
movie_id short_title                                            
787      Gate of Heavenly Peace, The                     3   5.0
3233     Smashing Time                                   2   5.0
989      Schlafes Bruder (Brother of Sleep)              1   5.0
1830     Follow the Bitch                                1   5.0
3172     Ulysses (Ulisse)                                1   5.0
...                                                    ...   ...
3237     Kestrel's Eye (Falkens öga)                     1   1.0
3312     McCullochs, The                                 1   1.0
3376     Fantastic Night, The (La Nuit Fantastique)      1   1.0
3460     Hillbillys in a Haunted House                   1   1.0
3904     Uninvited Guest, An                             1   1.0

[3706 rows x 2 columns]

> Method 1

#### Sorting by counts and mean for the movie ratings
- This gives an output of highest number of counts and its mean

In [106]:
movie_id_group.sort_values(by=['count','mean'], ascending=False)

count      mean
movie_id short_title                                                    
2858     American Beauty                                  3428  4.317386
260      Star Wars: Episode IV - A New Hope               2991  4.453694
1196     Star Wars: Episode V - The Empire Strikes Back   2990  4.292977
1210     Star Wars: Episode VI - Return of the Jedi       2883  4.022893
480      Jurassic Park                                    2672  3.763847
...                                                        ...       ...
3237     Kestrel's Eye (Falkens öga)                         1  1.000000
3312     McCullochs, The                                     1  1.000000
3376     Fantastic Night, The (La Nuit Fantastique)          1  1.000000
3460     Hillbillys in a Haunted House                       1  1.000000
3904     Uninvited Guest, An                                 1  1.000000

[3706 rows x 2 columns]

> Method 2

In [107]:
movie_id_group.nlargest(5, 'count')

count      mean
movie_id short_title                                                    
2858     American Beauty                                  3428  4.317386
260      Star Wars: Episode IV - A New Hope               2991  4.453694
1196     Star Wars: Episode V - The Empire Strikes Back   2990  4.292977
1210     Star Wars: Episode VI - Return of the Jedi       2883  4.022893
480      Jurassic Park                                    2672  3.763847

- movie_id and its rating counts

> Method 3

In [108]:
count_group

movie_id
1       2077
2        701
3        478
4        170
5        296
        ... 
3948     862
3949     304
3950      54
3951      40
3952     388
Name: rating, Length: 3706, dtype: int64

- if we put a threshold for movies having rating count greater than 2500, we get the top 13  with  rating counts higher than 2500
- if we put a threshold for movies having rating count greater than 3000, we get the only movie_id which as the rating count more than 3000 which is 2858.
- by refering to the movie_id_group table , the movie_id number 2858 belongs to 
### American Beauty which has the movie rating at 4.317 with  a highest rating count of 2858 is the Highest rated movie

In [109]:
count_group = users_ratings_movies.groupby("movie_id").count()["rating"]
movie_list = count_group[count_group > 2500].index.values
movie_list

array([ 260,  480,  589,  593,  608, 1196, 1198, 1210, 1270, 1580, 2028,
       2571, 2858], dtype=int64)

In [110]:
count_group = users_ratings_movies.groupby("movie_id").count()["rating"]
movie_list = count_group[count_group > 3000].index.values
movie_list

array([2858], dtype=int64)

> Method 4

### threshold is movies rating counts more than 2500 
- Sort the ratings counts in descending order

In [111]:
highest_rated = movie_id_group[movie_id_group['count'] > 2500] 
highest_rated .sort_values(by=['count'], ascending=False)

count      mean
movie_id short_title                                                    
2858     American Beauty                                  3428  4.317386
260      Star Wars: Episode IV - A New Hope               2991  4.453694
1196     Star Wars: Episode V - The Empire Strikes Back   2990  4.292977
1210     Star Wars: Episode VI - Return of the Jedi       2883  4.022893
480      Jurassic Park                                    2672  3.763847
2028     Saving Private Ryan                              2653  4.337354
589      Terminator 2: Judgment Day                       2649  4.058513
2571     Matrix, The                                      2590  4.315830
1270     Back to the Future                               2583  3.990321
593      Silence of the Lambs, The                        2578  4.351823
1580     Men in Black                                     2538  3.739953
1198     Raiders of the Lost Ark                          2514  4.477725
608      Fargo                                            2513  4.254676

### The Highest Rated Movie would be 
> The movie that has the highest rating count and the highest mean rating
### American Beauty is the highest rated movie with 3428 ratings count and 4.317 rating mean 

# What is a good rated movie for date night? (60 pts)

- Hint - highly rated movie by 
    - both partners (might be the same gender or not),
    - based on genre preferences,
    - age group can also be combined